src/Zio/FileSystems/ZipArchiveFileSystem.cs

```diff
@@ -0,0 +1,1034 @@
+                    return _archive.GetEntry(path);
+                }
+
+                return _entries.TryGetValue(path.ToLowerInvariant(), out var foundEntry) ? foundEntry : null;
+            }
+            finally
+            {
+                _entriesLock.ExitReadLock();
+            }
+        }
+
+        /// <inheritdoc />
+        protected override UPath ConvertPathFromInternalImpl(string innerPath)
+        {
+            return new UPath(innerPath);
+        }
+
+        /// <inheritdoc />
+        protected override string ConvertPathToInternalImpl(UPath path)
+        {
+            return path.FullName;
+        }
+
+        /// <inheritdoc />
+        protected override void CopyFileImpl(UPath srcPath, UPath destPath, bool overwrite)
+        {
+            if (srcPath == destPath)
+            {
+                throw new IOException("Source and destination path must be different.");
+            }
+
+            var entry = GetEntry(srcPath.FullName);
+            if (entry == null)
+            {
+                if (DirectoryExistsImpl(srcPath))
+                {
+                    throw new UnauthorizedAccessException(nameof(srcPath) + " is a directory.");
+                }
+
+                if (!DirectoryExistsImpl(srcPath.GetDirectory()))
+                {
+                    throw new DirectoryNotFoundException(srcPath.GetDirectory().FullName);
+                }
+
+                throw FileSystemExceptionHelper.NewFileNotFoundException(srcPath);
+            }
+
+            var parentDirectory = destPath.GetDirectory();
+            if (!DirectoryExistsImpl(parentDirectory))
+            {
+                throw FileSystemExceptionHelper.NewDirectoryNotFoundException(parentDirectory);
+            }
+
+            if (DirectoryExistsImpl(destPath))
+            {
+                if (!FileExistsImpl(destPath))
+                {
+                    throw new IOException("Destination path is a directory");
+                }
+            }
+
+            var destEntry = GetEntry(destPath.FullName);
+            if (destEntry != null)
+            {
+#if NETSTANDARD2_1_OR_GREATER
+                if ((destEntry.ExternalAttributes & (int)FileAttributes.ReadOnly) == (int)FileAttributes.ReadOnly)
+                {
+                    throw new UnauthorizedAccessException("Destination file is read only");
+                }
+#endif
+                if (!overwrite)
+                {
+                    throw FileSystemExceptionHelper.NewDestinationFileExistException(srcPath);
+                }
+
+                RemoveEntry(destEntry);
+                TryGetDispatcher()?.RaiseDeleted(destPath);
+            }
+
+            destEntry = CreateEntry(destPath.FullName);
+            using (var destStream = destEntry.Open())
+            {
+                using var srcStream = entry == null ? new FileStream(srcPath.FullName, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite) : entry.Open();
```

Comment:
`entry` cannot be null because there is a throw above, right?

Reply:
right, changed it